File: Ant/chapter2/cpp/counting_prime_in_a_segment.cpp
#include <fstream>
#include <stdio.h>

#define MAX_L 1'000'000
#define MAX_SQRT_B 1'000'000

using namespace std;

typedef long long ll;

ll a, b;
bool is_prime_small[MAX_SQRT_B];
bool is_prime[MAX_L];

void segment_sieve(ll a, ll b) {
  for (ll i = 0; i * i < b; ++i) {
    is_prime_small[i] = true;
  }
  for (ll i = 0; i < b - a; ++i) {
    is_prime[i] = true;
  }
  for (ll i = 2; i * i < b; ++i) {
    if (is_prime_small[i]) {
      for (ll j = i * 2; j * j < b; j += i) {
        is_prime_small[j] = false;
      }
      // ((a + i - 1) / i) * i = the smallest multiple of i which is more than or equal to a
      for (ll j = max(2LL, (a + i - 1) / i) * i; j < b; j += i) {
        is_prime[j - a] = false;
      }
    }
  }
}

int main() {
  ifstream ifs("../testset/counting_prime_in_a_segment/test2.txt");
  ifs >> a >> b;
  segment_sieve(a, b);
  int res = 0;
  for (ll i = 0; i < b - a; ++i) {
    res += is_prime[i];
  }
  printf("%d\n", res);
}

File: Ant/chapter2/python/counting_prime_in_a_segment.py
def ns(f):
    return next(f).strip()


with open("../testset/counting_prime_in_a_segment/test2.txt", 'r') as f:
    a, b = map(int, ns(f).split())


_b = int(b ** 0.5)
is_prime_small = [True] * _b
is_prime = [True] * (b - a)


def segment_sieve(a, b):
    global is_prime_small, is_prime
    for i in range(2, _b):
        if is_prime_small[i]:
            for j in range(i * 2, _b, i):
                is_prime_small[j] = False
            for j in range(max(2, (a + i - 1) // i * i), b, i):
                is_prime[j - a] = False


segment_sieve(a, b)
print(len(list(filter(lambda x: x, is_prime))))
